Docs: Overhaul Okta SAML guide. (#23053)
gravitational/teleport: docs/pages/access-controls/sso/okta.mdx

---
title: Authentication With Okta as an SSO provider
description: How to configure Teleport access using Okta for SSO
videoBanner: SM4Am-i8cj4
---

## How to use Okta as a single sign-on (SSO) provider for SSH

This guide covers how to configure [Okta](https://www.okta.com/) to provide
single sign on (SSO) identities to Teleport Enterprise and Teleport Enterprise
Cloud. When used in combination with role based access control (RBAC), this allows
Teleport administrators to define policies like:

- Only members of "DBA" group can access into PostgreSQL databases.
- Developers must never SSH into production servers.
- Members of the HR group can view audit logs but not production environments.

## Prerequisites

- An Okta account with admin access. Your account must include users and at
  least two groups.

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- A Teleport role with access to edit and maintain `saml` resources. This is
  available in the default `editor` role.

(!docs/pages/includes/tctl.mdx!)

(!docs/pages/includes/enterprise/samlauthentication.mdx!)

> ## Step 1/4. Create & assign groups

Author: I've added some explanation here, leaving unresolved for you to 👍 or 👎  it.

### Create Groups

Create two groups: "okta-dev" and "okta-admin". In the Okta dashboard go to the

> navigation bar and click **Directory** -> Groups, then **Add group**:

Author: @ptgott I originally formatted it this way because "Directory" is bold but groups is not in the Okta nav:

![image](https://user-images.githubusercontent.com/2349184/226625870-54de2e04-a9a8-4091-a3b7-32bdcb836f58.png)

Do you think We should align with external system formatting, or use consistent bold to indicated each nav item in a process?

> navigation bar and click **Directory** -> Groups, then **Add group**:

Reviewer: I think using bold consistently helps set the nav flow apart from the main text regardless of how the menu is formatted

![Create Group Devs](../../../img/sso/okta/okta-saml-2.1@2x.png)

Repeat for the admin group:

![Create Group Devs](../../../img/sso/okta/okta-saml-2.2.png)

> ## Step 2/4. Configure Okta

Author: I've added some explanation here, leaving unresolved for you to 👍 or 👎  it.

First, create a SAML 2.0 Web App in Okta:

> From the main navigation menu, select **Applications** -> Applications, and click

Author: Same note as above.

**Create App Integration**. Select SAML 2.0, then click **Next**.

![Create APP](../../../img/sso/okta/okta-saml-1.png)

On the next screen (**General Settings**), provide a name and optional logo for your new app, then
click **Next**. This will bring you to the **Configure SAML** section.

### Configure the App

Provide the following values to their respective fields:

#### General

- Single sign on URL: `https://<cluster-url>/v1/webapi/saml/acs/new_saml_connector`
- Audience URI (SP Entity ID): <nobr>`https://<cluster-url>/v1/webapi/saml/acs/new_saml_connector`</nobr>
- Name ID format `EmailAddress`
- Application username `Okta username`

> Replace `<cluster-url>` with your Teleport proxy address or cloud tenant (e.g. `mytenant.teleport.sh`).

Author: I use Okta for my home cluster's IdP without defining a port in the Okta Application setting. You can define one, but since the Teleport Proxy service can't be configured to listen on alternate ports (at least since the last time I asked), It doesn't matter if you do or don't include `:443`.

#### Attribute Statements

- Name: `username`  | Name format: `Unspecified` | Value: `user.login`

#### Group Attribute Statements

We will map our Okta groups to the SAML Attribute statements (special signed
metadata exposed via a SAML XML response).

- Name: `groups` | Name format: `Unspecified`
- Filter: `Matches regex` |  `.*`

The configuration page should now look like this:

![Configure APP](../../../img/sso/okta/setup-redirection.png)

<Admonition type="warning" >
RegEx requires `.*`
</Admonition>

<Admonition type="tip">

> Notice that we have set "NameID" to the email format and mapped the groups with

Reviewer: It would be good to have some information as to why the values we mention in this `Admonition` are especially important. Is there a reason we tell the reader about these and not the other values we mention in the instructions above?

Author: I've added some explanation here, leaving unresolved for you to 👍 or 👎  it.

a wildcard regex in the Group Attribute statements. We have also set the "Audience"
and SSO URL to the same value.
</Admonition>

Once you've filled the required fields, click **Next**, then finish the app creation wizard.

From the **Assignments** tab of the new application page, click **Assign**.
Assign the newly created groups access to the app.

![Configure APP](../../../img/sso/okta/okta-saml-3.1.png)

> ### Save IdP metadata

Reviewer: As above, it would be nice to have a sentence about why were are saving IdP metadata.

Author: I've added some explanation here, leaving unresolved for you to 👍 or 👎  it.

From the app's **Sign On** tab, scroll down to **SAML Signing Certificates**.
Click **Actions** for the SHA-2 entry, then "View IdP metadata":

![View Okta IdP Metadata](../../../img/sso/okta/okta-copy-metadata.png)

Copy the URL to the metadata file for use in our Teleport configuration.

<Admonition type="tip">
You can also right click on the "View IdP metadata" link and select
"Copy Link" or "Copy Link Address".
</Admonition>

## Step 3/4. Create a SAML connector

> Define a SAML connector [resource](../../reference/resources.mdx) in a local

Reviewer: Do we need this link here?

file named `okta-connector.yaml`:

```yaml
(!examples/resources/saml-connector.yaml!)
```

> Update the value of `acs` with your Teleport Proxy address or cloud tenant

Reviewer: ```suggestion
Update the value of `acs` with your Teleport Proxy address or Enterprise Cloud tenant
```

> (e.g. `mytenant.teleport.sh`), and replace the value of `entity_descriptor_url`

Reviewer: Should there be a port here as well?

Author: Nope.

---
title: Authentication With Okta as an SSO Provider
description: How to configure Teleport access using Okta for SSO
videoBanner: SM4Am-i8cj4
---

This guide covers how to configure [Okta](https://www.okta.com/) to provide
single sign on (SSO) identities to Teleport Enterprise and Teleport Enterprise
Cloud. When used in combination with role-based access control (RBAC), this allows
Teleport administrators to define policies like:

- Only members of the "DBA" group can access PostgreSQL databases.
- Developers must never SSH into production servers.
- Members of the HR group can view audit logs but not production environments.

## Prerequisites

- An Okta account with admin access. Your account must include users and at
  least two groups. If you don't already have Okta groups you want to assign to
  Teleport roles don't worry, we'll create example groups below.

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- A Teleport role with access to edit and maintain `saml` resources. This is
  available in the default `editor` role.

(!docs/pages/includes/tctl.mdx!)

(!docs/pages/includes/enterprise/samlauthentication.mdx!)

## Step 1/4. Create & assign groups

> Okta provides group membership as a SAML assertion in the user data it provides

Reviewer: ```suggestion
Okta indicates a user's group membership as a SAML assertion in the data it provides
```
To prevent repeating "provides"

> to Teleport. We will configure Teleport to assign roles based on these groups.

Reviewer: ```suggestion
to Teleport. We will configure Teleport to assign roles (called "dev" and "admin") based on these groups.
```
So it's clear why we're referring to "dev" and "admin" in the next paragraph

If you already have Okta groups you want to assign to "dev" and "admin" roles in
Teleport, you can skip to the [next step](#step-24-configure-okta).

### Create Groups

Create two groups: "okta-dev" and "okta-admin". In the Okta dashboard go to the
navigation bar and click **Directory** -> Groups, then **Add group**:

![Create Group Devs](../../../img/sso/okta/okta-saml-2.1@2x.png)

Repeat for the admin group:

![Create Group Devs](../../../img/sso/okta/okta-saml-2.2.png)

## Step 2/4. Configure Okta

> In this section we will create an application in the Okta dashboard to define
> and allow our Teleport cluster to access Okta as an IdP provider. We'll also
> locate the address where Okta provides their IdP metadata to provide to Teleport.

Reviewer: ```suggestion
In this section we will create an application in the Okta dashboard to allow our Teleport cluster to access Okta as an IdP provider. We'll also
locate the address that Okta uses to provides their IdP metadata to Teleport.
```
To tighten this up a bit and avoid repeating "provides"

- Single sign on URL: `https://<cluster-url>/v1/webapi/saml/acs/new_saml_connector`
- Audience URI (SP Entity ID): <nobr>`https://<cluster-url>/v1/webapi/saml/acs/new_saml_connector`</nobr>
- Name ID format `EmailAddress`
- Application username `Okta username`

Replace `<cluster-url>` with your Teleport Proxy Service address or Enterprise
Cloud tenant (e.g. `mytenant.teleport.sh`).

#### Attribute Statements

- Name: `username`  | Name format: `Unspecified` | Value: `user.login`

#### Group Attribute Statements

We will map our Okta groups to SAML attribute statements (special signed metadata
exposed via a SAML XML response), so that Teleport can discover a user's group
membership and assign matching roles.

- Name: `groups` | Name format: `Unspecified`
- Filter: `Matches regex` |  `.*`

The configuration page should now look like this:

![Configure APP](../../../img/sso/okta/setup-redirection.png)

<Admonition type="warning" >
The "Matches regex" filter requires the literal string `.*` in order to match all
content from the group attribute statement.
</Admonition>

<Admonition type="tip">
Notice that we have set "NameID" to the email format and mapped the groups with

> a wildcard regex in the Group Attribute statements. We have also set the "Audience"
> and SSO URL to the same value. This is so Teleport can read and use Okta users'
> email address to create their username in Teleport, instead of relying on additional
> name fields.

Reviewer: ```suggestion
a wildcard regex in the Group Attribute statements. We have also set the "Audience"
and SSO URLs to the same value. This is so Teleport can read and use Okta users'
email addresses to create their usernames in Teleport, instead of relying on additional
name fields.
```

$ tctl create okta-connector.yaml
```

## Step 4/4. Create a developer Teleport role

Now let's create a new role to pull in external information from Okta.
Create the local file `dev.yaml` with the content below.

> Notice the `{{external.username}}` login. It configures Teleport to look at the

Reviewer: Should this paragraph go after the code snippet so the connection between "the content below" and the snippet to copy is more obvious?

```

- Devs are only allowed to login to nodes labelled with `access: relaxed` label.

> - Developers can log in as `ubuntu` user or a user matching their email prefix.

Reviewer: Is there a way to avoid repeating "Devs"/"Developers" in this list?

Author: I reformatted it a bit, but it could still be considered redundant. Happy to take any suggestions to further improve this section.

- Developers do not have any "allow rules" i.e. they will not be able to
  see/replay past sessions or re-configure the Teleport cluster.

> Use tctl to create this role on the auth server:

Reviewer: ```suggestion
Use tctl to create this role on the Auth Service:
```

Reviewer: Also, does this need to be on the Auth Service, or can we use tctl remotely here?

Author: This is saying that the role exists in the Auth Service after creation, not that `tctl` is run on the Auth Service host specifically. I'll see if I can rephrase.
